### PR TITLE
fix: add support for float16 metrics serialization

### PR DESCRIPTION
**Pull request by BlueskyFR**
## Description

Adds support for float16 metrics serialization

## Commentary (optional)

When running an experiment using keras while having enable mixed precision using `tf.keras.mixed_precision.set_global_policy("mixed_float16")` (TF 2.6), the metrics are serialized before being sent by socket (to the master I guess?).
However, while float64 and float32 are supported, **float16** is not one of them: this patch hopefully fixes it.

## Checklist

- [X] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
See [Release Note](../docs/release-notes/README.md) for details.
- [X] Licenses should be included for new code which was copied and/or modified from any external code.

[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234

### CODE REVIEW

**Review by rb-determined-ai — APPROVED** [review]
changes look good, thank you for the fix!

**Comment by BlueskyFR** [comment]
I was not yet able to test it locally though, so I am wondering if anything else will need to be changed too

**Comment by rb-determined-ai** [comment]
Ok.  If you are not able to test it manually by the time we work out the CLA details, I'll write a test before we merge this.

**Comment by BlueskyFR** [comment]
> Ok. If you are not able to test it manually by the time we work out the CLA details, I'll write a test before we merge this.

Is there some build instructions somewhere so that I can test the fixed agent locally?

**Comment by BlueskyFR** [comment]
@cla-bot[bot] check

**Comment by rb-determined-ai** [comment]
> Is there some build instructions somewhere so that I can test the fixed agent image locally?

Ah, I guess not.  Actually our build system is a bit of a pain.  You are welcome to clone the repo and build it `make get-deps; make build`, but you'll need working go environment, protoc, java, python, sphinx, node... it's a bit unwieldy.

On the other hand, for simple fixes like this you can hotpatch the python harness in your live cluster in a `startup-hook.sh` that you put in your model directory:
```sh
file="$(python -c 'import determined.util as x; print(x.__file__)')"
sed -i -e '/if isinstance(obj, np.float32):/a \            return float(obj)\n        if isinstance(obj, np.float16):'  "$file"
```

That will literally just modify the source of the installed python library in the container just before running your code.

> @cla-bot[bot] check

We're still talking to legal about what exactly required in your case.  At any rate, the next release is next week so there's not a huge rush here.

**Comment by BlueskyFR** [comment]
I'll run some tests tomorrow then :)
Once you get any feedback on the cla make sure to keep me informed ;)

**Comment by rb-determined-ai** [comment]
@cla-bot[bot] check

**Comment by rb-determined-ai** [comment]
@BlueskyFR we talked to legal and you are good to go!

**Comment by rb-determined-ai** [comment]
@BlueskyFR have you had a chance to test this yet?  In either case, please mark the PR as ready for review.

**Comment by rb-determined-ai** [comment]
Not sure where you went but I'll just land this.

**Comment by BlueskyFR** [comment]
> Not sure where you went but I'll just land this.

Hi @rb-determined-ai, sorry for this: I returned to school so I didn't see the Github notifications :(
Glad to see you merged it! :D
